SEP-0010: Add Memo Support (#726)
stellar/stellar-protocol: ecosystem/sep-0010.md

This SEP defines the standard way for clients such as wallets or exchanges to create authenticated web sessions on behalf of a user who holds a Stellar account. A wallet may want to authenticate with any web service which requires a Stellar account ownership verification, for example, to upload KYC information to an anchor in an authenticated way as described in [SEP-12](sep-0012.md).

> This SEP also supports clients who control [pooled accounts](#pooled-accounts) containing funds owned by _many_ users. These clients manage their users' Stellar accounts and hold the secret keys for those accounts.

Reviewer: Suggested change to align with the other comment about pooled accounts:
```suggestion
This SEP also supports authenticating clients who will prove a claim to a [memo](#memos) value of a Stellar account that is useful for authenticating clients who share a Stellar account.
```

Author: `who will prove a claim to a memo value` doesn't read well. What about:
```suggestion
This SEP also supports authenticating clients that share a Stellar account. Each unique client should be identified with a [memo](#memos) value included in the initial GET request as well as the JWT body.
```

Reviewer: That's clearer. I think we should leave out technical details in this paragraph, and instead of stating what a client should do, we don't need to be prescriptive, and can highlight who it relates to. 
```suggestion
This SEP also supports authenticating clients that share a Stellar account and use [memos](#memos) to distinguish users or accounts for payments.
```

> This SEP also supports clients who control [pooled accounts](#pooled-accounts) containing funds owned by _many_ users. These clients are _custodians_ of their users' Stellar accounts and possess the secret keys for those accounts.

Reviewer: I think we should avoid the use custodian. That term has specific legal meaning which is outside the scope of this technical spec. I also think we're being unnecessarily prescriptive to discuss secret keys of accounts.
```suggestion
This SEP also supports clients who control [pooled accounts](#pooled-accounts) containing funds owned by _many_ users.
```

----------|---------------|------------
`account` | `G...` string | The stellar account that the wallet wishes to authenticate with the server
`home_domain` | string | (optional) The [fully qualified domain name](https://en.wikipedia.org/wiki/Fully_qualified_domain_name) of the service requiring authentication. SEP-10 signing servers used to generate tokens for multiple different web services can use this parameter to create valid tokens for each service. If not provided by the client, a default should be used.

> `memo` | integer | (optional) The 64-bit unsigned integer used to identify the user in the case of a pooled `account`. Note that the `memo_type` field is not needed because all values are of type `MEMO_ID`.

Reviewer: See note further down about leveraging the `M` address format to encode the memo and account together. I think we could leverage that format for just SEP-10.

}
```

> ### Pooled Accounts
> 
> In the context of Stellar, pooled accounts are Stellar accounts that are held and managed by an entity that is _not_ the owner of the funds within the account. An alternative approach to pooled accounts that still allows the client entity to manage and hold account secret keys is to create a separate Stellar account for each unique user. While this is possible, a much more common practice is to hold the funds of many users in a single _pooled_ account.
> 
> The `memo` parameter supported in `GET <WEB_AUTH_ENDPOINT>` API calls can be used to create `account`-`memo` composite identifiers for the web service's users. The decoded contents of the token returned from `POST <WEB_AUTH_ENDPOINT>` calls should contain both the `account` and `memo` values so the service requiring SEP-10 authentication can tie each request to the user.
> 
> In the future, there may be a more streamlined way to specify users within pooled accounts. This section and the API specification will be updated when that change happens.

Reviewer: I think this section needs to be used to the bottom of the document because it is informational discussion. Right now it is sitting in between the challenge and token sections which should be back to back.

I also think we should remove the discussion about fund ownership, secret keys, unique user's, the new concept of "pooled accounts", and the theories about what might happen in the future. I think we should call out in this section why memos are relevant and when they can be used. This aligns more specifically with the raw constructs.

```suggestion
### Memos

Stellar transactions may be specified with a memo and services that share a single account for many users typically use the memo as a method to distinguishing one user from another, or one internal off-network account from another.

The `memo_type` and `memo` parameters supported in `GET <WEB_AUTH_ENDPOINT>` API calls can be used to communicate to the server that the client intends to authenticate a Stellar account with an additional claim that the authenticating user is the provided `memo`.
```

Author: I think its important to have a section header that hints at what it enables. We don't need to strictly defined Pooled Accounts, but a person looking for how a pooled account is authenticated will appreciate the more direct label.

Reviewer: Multi-user accounts is ambiguous with multi-signer, which is a different. I think some of the issues with names like multi-user, pooled, custodial, shared, etc is that they are not terms used by anyone else. In our other SEPs that ellude to the same use cases we discuss only memos, e.g. SEP-2. For this reason I think titling this as memos is sufficient, especially with the link from the summary.

Reviewer: Also if muxed addresses take off ultimately it will become the term people use and this current functionality will fade into the background.

}
```

### Memos

Stellar transactions may be specified with a memo and services that share a single account for many users typically use the memo as a method to distinguishing one user from another, or one internal off-network account from another.

The `memo_type` and `memo` parameters supported in `GET <WEB_AUTH_ENDPOINT>` API calls can be used to communicate to the server that the client intends to authenticate a Stellar account with an additional claim that the authenticating user has control of or is possession of the provided `memo`.

> 

Reviewer: Can we move this to the bottom of the document after the `JWT Expiration` section?

}
```

### Pooled Accounts

> Before defining pooled accounts, we must first define custodial accounts. In the context of Stellar, custodial accounts are Stellar accounts that are held and managed by an entity that is _not_ the owner of the funds within the account. Specifically, custodial entities hold the secret key(s) of the accounts they control. These custodial entities can create individual accounts for each user, but a much more common practice is to hold the funds of many users in a single _pooled_ account.

Reviewer: I don't think we can use the term custodial here. I believe it is a legal term, which means it could be interpreted differently in different regions of the world, and we maybe don't want to get into that in a technical specification. Can we define what we need to define here without using the term custodial?

Author: Is there a better term you can think of?

Author: Because not all custodial accounts are pooled accounts, so we need a synonym or we need to define a term here in the SEP.

Reviewer: I think we can start with "Pooled accounts are..." and define what they are without referencing custodial accounts. What drives the need to discuss custodial?

Author: What do we gain from using the M-address format? Memos are the tried and true solution to this problem without M-addresses, and without full `M`-address support in Horizon/SDK's I don't see a reason to use them.

* `iss` (the principal that issued a token, [RFC7519, Section 4.1.1](https://tools.ietf.org/html/rfc7519#section-4.1.1)) — a [Uniform Resource Identifier (URI)] for the issuer (`https://example.com` or `https://example.com/G...`)
* `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account _potentially_ contcatenated with the `memo` value originally specified by the client using an `-` character
  * For example, when `account` is pooled, the user can be identified using the 32-byte hash memo: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`
  * The `sub` field without a `memo` value would be a plain Stellar public key: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4`
  * Web services that are passed a JWT token can use the `sub` memo value to internally identify the user who owns the funds held by the client on each request
* `iat` (the time at which the JWT was issued [RFC7519, Section 4.1.6](https://tools.ietf.org/html/rfc7519#section-4.1.6)) — current timestamp (`1530644093`)

> * `exp` (the expiration time on or after which the JWT must not be accepted for processing, [RFC7519, Section 4.1.4](https://tools.ietf.org/html/rfc7519#section-4.1.4)) — a server can pick its own expiration period for the token (`1530730493`)

Author: I agree, this is the route to go. It wouldn't surprise me if wallet-anchor services are already adding non-standard fields to their JWT's.

* use operations's source account to determine the authenticating client and perform any additional service-specific validations.
* if the transaction has a memo:
  * verify the transaction's memo and source account can be used to internally identify the authenticated user;

The verification process confirms that a user holds an account. Depending on your application this may mean complete signing authority, some threshold of control, or being a signer of the account. See [Verification](#verification) for examples.

Upon successful verification, service responds with a session JWT, containing the following claims:

* `iss` (the principal that issued a token, [RFC7519, Section 4.1.1](https://tools.ietf.org/html/rfc7519#section-4.1.1)) — a [Uniform Resource Identifier (URI)] for the issuer (`https://example.com` or `https://example.com/G...`)
* `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account potentially contcatenated with the `memo` value originally specified by the client using an `-` character

>   * For example, when `account` is pooled, the user can be identified using the 32-byte hash memo: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`

Reviewer: I agree we don't know yet if there will be more effort immediately put into rolling out multiplexed accounts in all areas of the SDKs, Horizon, etc, however I think if we are willing to limit pooled account memo support here to int64s then we may as well leverage [SEP-23](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0023.md) and use the `M` strkey format for a multiplexed address as the format that we encode the account and the memo ID.

If multiplexed addresses get their moment then we'll already be using the appropriate address format here.

In the event multiplexed addresses never take off in other ways using the `M` strkey encoding here won't be any worse than us creating our own format.

By using the `M` strkey format we avoid adding a new format here that is unique to SEP-10.

@JakeUrban @tomerweller @ire-and-curses Thoughts?

Author: I'm concerned the presence of `M` address in SEP-10 will imply wider ecosystem support. Won't developers be confused to see a new type of Stellar account without being able to query Horizon for its `/account` JSON? I don't want to conflate the _need_ to encode a memo with the _desire_ to use `M` addresses.

Reviewer: I think the fact the M accounts cannot be queried via Horizon endpoints is what allows us to use it in a limited capacity. It will be no different if someone tries to query Horizon with `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`, they'll see the same error.

Author: What do we gain from using the M-address format? Memos are the tried and true solution to this problem without M-addresses, and without full M-address support in Horizon/SDK's I don't see a reason to use them.

Author: I can also see future issues where the `M`-addresses used here complicate the process for full `M`-address support. For example, how will the `memo` value be encoded into the `G`-address to get an `M`-address? It will likely be a different algorithm than the algorithm used to generate real `M`-addresses.

>   * For example, when `account` is pooled, the user can be identified using the 32-byte hash memo: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`

Reviewer: I think we need two examples here, one without a pooled account and one with, so that it is clearer memo is optional.

  * verify the transaction's memo and source account can be used to internally identify the authenticated user;

The verification process confirms that a user holds an account. Depending on your application this may mean complete signing authority, some threshold of control, or being a signer of the account. See [Verification](#verification) for examples.

Upon successful verification, service responds with a session JWT, containing the following claims:

* `iss` (the principal that issued a token, [RFC7519, Section 4.1.1](https://tools.ietf.org/html/rfc7519#section-4.1.1)) — a [Uniform Resource Identifier (URI)] for the issuer (`https://example.com` or `https://example.com/G...`)
* `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account potentially contcatenated with the `memo` value originally specified by the client using an `-` character
  * For example, when `account` is pooled, the user can be identified using the 32-byte hash memo: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`

>   * Web services that are passed a JWT token can use the `sub` memo value to internally identify the user who owns the funds held by the client on each request

Reviewer: We need to define how each memo type is formatted explicitly.

Author: This is resolved now that memos are only ID's

>   * Web services that are passed a JWT token can use the `sub` memo value to internally identify the user who owns the funds held by the client on each request

Reviewer: Unfortunately the `sub` field must be a string limiting us to the delimiter approach, but we do not need to use `sub`. This could be a good time to make this a breaking change where by we abandon the use of the `sub` field and define new fields like `account` and `memo`. Something like:

Standard account:
```
"account": "G..."
```

Pooled or multiplexed account:
```
"account": "M..."
```
or
```
"account": {
    "account": "G...",
    "memo": "1234"
}
```

Author: I'm usually not one to back down from a breaking change but SEP-10 is one of the most widely used standards in the Stellar ecosystem, and given the current state of SEP-10 regarding the 2.0 SDK updates, I think we should try not to break things until we define a better transition process.

Reviewer: Yeah that's a good point.

I worry that delimiting the data in `sub` will come back to bite us in ways we don't expect, but I don't have any concrete examples.

I like that with this proposal the SDKs we know of should break if they see the hyphen and haven't been upgraded to support a memo. So they shouldn't accidentally validate a JWT as proving control of an account when it only proves alignment with a memo value.

Some other options:
- We could also define the `sub` to be encoded in some other format, such as JSON, so it is JSON stringified. This would make it predictable using common technology to build and parse. The value of sub would become:
   ```
   "sub": "{\"acc\":\"G...\",\"memo\":\"1234\"}"
   ```
   This makes it clear for any developer looking at the data at what they have because it is self-describing.
- We could use the strkey multiplexed format in SEP-10 without supporting them generally in other ways. This would allow us to adopt SEP-10 in the direction of that standard without waiting for any other part of it being implemented. We'd only use it to encode the data we have. This would also limit the memo types to the 64-bit ID.

Author: I'm not familiar with the strkey multiplexed format I'll take a look at it and come back with comments/questions.

Regarding the first 'other' option, it would certainly work, but I think it may confuse clients. If they've already deserialized the JSON string from the response body, it seems strange to deserialize a specific attribute within the JWT.

Personally I think the current design is the way to go. The one concern I have is that the current design doesn't provide the `memo_type` value in the `sub` field of the JWT. I think we need a better format for encoding this info.

The SDK's would be updated to validate this format if present, otherwise the original Stellar public key value would be expected.

Reviewer: We can avoid needing to include the memo type if we limit the memo types allowed to the memo types allowed by multiplexed addresses / muxed accounts, which is the memo id type, a 64-bit integer. If we're thinking ahead to when multiplexed addresses are in use we'll be limited to those memo types anyway. My understanding is the research @stanford-scs did highlighted that most folks only needed that memo type. @tomerweller Thoughts?

Author: @tomerweller bump, starting to think about this again

Reviewer 2: Yes. We discussed this during last year's meridian protocol session and there was a wide agreement that int64s are enough. Which will make transitioning to muxed accounts easier.

The client can examine the `network_passphrase` (if defined) that the server includes in its response from the challenge endpoint to be sure it's using the correct passphrase and is connecting to the server that it expected.

> ## JWT best practices

Reviewer: Could we do the JWT expiration/best practices change in a separate PR? It seems unrelated to the rest of this change.

Author: It is unrelated but not meaningful, just combines the two JWT sections since the `JWT best practices` was one sentence.

Author: I can undo the change if we don't want to include it
